source/blog/_posts/2015-07-15-cross-cutting.md

```diff
@@ -0,0 +1,383 @@
+    // logging $event->getItemId;
+});
+
+$container->get('event_manager')->subscribe('cart.afterPurchase', function(AfterPurchaseEvent $event) {
+    // logging $event->getItemId;
+});
+```
+
+A very common PubSub implementation is the [symfony event dispatcher](https://github.com/symfony/EventDispatcher/blob/master/EventDispatcherInterface.php)
+which also comes along with a handy [EventSubscriberInterface](https://github.com/symfony/EventDispatcher/blob/master/EventSubscriberInterface.php)
+which makes subscribing events even easier and is also available in Shopware.
+
+  
+### Conclusion
+Both approaches will allow you to remove the direct dependency to logging, transactions and ACL from the cart implementation
+and address those issues in specific observers / subscribers. 
+
+Both patterns tend to hide the actually executed code of the `Cart` class a bit, as the dependencies are registered during runtime
+and cannot be statically inspected. For that reason, you cannot tell beforehands, if e.g. only the `Logger` service is subscribed
+or also the ACL service. This kind of abstraction can make debugging harder.
+ 
+Another common problem is the handling of context information: In the observer example a simple `Array` was used, which is convenient
+but might lead to issues regarding typos and debuggability. This might easily be changed to event objects, as
+the PubSub example shows: The classes `BeforePurchaseEvent` and `AfterPurchaseEvent` define a clean interface for a specific
+event and can easily inspected for debugging reasons.
+I already mentioned, that the PHP [SplObserver](http://php.net/manual/de/class.splobserver.php) interface for an observer pattern
+defines, that a reference of the subject (the `Cart` class in our case) is passed. From my experience with the Shopware
+event system, I would consider this a bad practice, as this will allow the observers to hardly bind to the subject, even though
+this might not be necessary at all. Selectively passing relevant context information also will make sure, that the observer
+ do not modify e.g. internal state of the passed class. 
+ 
+This is related to a third issue I'd like to address: In many cases some kind of "backchannel" is wanted, so that a 
+event listener can e.g. filter some context variables (`filter` event in Shopware) or return objects (`notifyUntil`).
+This might be considered bad practice, too, as multiple event listener changing the same event object by reference will lead
+to hard to debug errors and side effects in some cases. Speaking of [loose coupling](https://en.wikipedia.org/wiki/Loose_coupling)
+as one of the main goals for our event system, by-reference changes of context objects and behavioural decisions depending
+on the return value of an event, are not ideal but hard to avoid. 
+
+## Command Bus
+The command bus is based on the [command pattern](https://en.wikipedia.org/wiki/Command_pattern) and also allows you to address
```

Comment:
The command pattern does not define a command bus. A command as defined in the command pattern is self executable.